src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp

```diff
@@ -0,0 +1,129 @@
+ *                                                                         *
+ *   You should have received a copy of the GNU Library General Public     *
+ *   License along with this library; see the file COPYING.LIB. If not,    *
+ *   write to the Free Software Foundation, Inc., 59 Temple Place,         *
+ *   Suite 330, Boston, MA  02111-1307, USA                                *
+ *                                                                         *
+ ***************************************************************************/
+
+
+#include "PreCompiled.h"
+
+#ifndef _PreComp_
+# include <BRepAdaptor_Curve.hxx>
+# include <BRepAdaptor_Surface.hxx>
+# include <Geom_Line.hxx>
+# include <Geom_Plane.hxx>
+# include <Precision.hxx>
+
+# include <QAction>
+# include <QKeyEvent>
+# include <QMessageBox>
+# include <QRegExp>
+# include <QTextStream>
+
+# include <TopoDS.hxx>
+# include <gp_Ax1.hxx>
+# include <gp_Lin.hxx>
+# include <gp_Pln.hxx>
+# include <sstream>
+#endif
+
+#include "TaskFemConstraintOnBoundary.h"
+#include <App/Application.h>
+#include <Base/Tools.h>
+#include <Gui/Command.h>
+#include <Gui/Selection.h>
+#include <Gui/SelectionFilter.h>
+
+
+using namespace FemGui;
+using namespace Gui;
+
+/* TRANSLATOR FemGui::TaskFemConstraintOnBoundary */
+
+TaskFemConstraintOnBoundary::TaskFemConstraintOnBoundary(ViewProviderFemConstraint *ConstraintView, QWidget *parent, const char* pixmapname)
+    : TaskFemConstraint(ConstraintView, parent, pixmapname)
+{
+}
+
+TaskFemConstraintOnBoundary::~TaskFemConstraintOnBoundary()
+{
+}
+
+void TaskFemConstraintOnBoundary::_addToSelection(bool checked)
+{
+    if (checked)
+    {
+        const auto& selection = Gui::Selection().getSelectionEx(); //gets vector of selected objects of active document
+        if (selection.empty()) {
+            this->clearButtons(refAdd);
+            selChangeMode = refAdd;
+        } else {
+            this->addToSelection();
+            clearButtons(none);
+        }
+    } else {
+        exitSelectionChangeMode();
+    }
+}
+
+void TaskFemConstraintOnBoundary::_removeFromSelection(bool checked)
+{
+    if (checked)
+    {
+        const auto& selection = Gui::Selection().getSelectionEx(); //gets vector of selected objects of active document
+        if (selection.empty()) {
+            this->clearButtons(refRemove);
+            selChangeMode = refRemove;
+        } else {
+            this->removeFromSelection();
+            clearButtons(none);
+        }
+    } else {
+        exitSelectionChangeMode();
+    }
+}
+
+void TaskFemConstraintOnBoundary::onSelectionChanged(const Gui::SelectionChanges&msg)
+{
+    if (msg.Type == Gui::SelectionChanges::AddSelection) {
+        switch (selChangeMode) {
+        case refAdd:
+            // TODO: Optimize to just perform actions on the newly selected item
+            this->addToSelection();
+            // ui->lw_references->addItem(makeRefText(msg.pObjectName, msg.pSubName));
```

Comment:
this comment can go

Reply:
I don't think so. The TODO is there to suggest that the selection changes can be used to reduce the computational effort, and the commented line suggests how to to do it.

Reply:
OK. However for an outsider, the commented code should follow the Todo to make this clear

Reply:
OK done!